### ISSUE TITLE
[Feat] Add a DefectDojo integration

### ISSUE THREAD

**Issue by fb33** [oss-review-toolkit/ort#2736]
Some ideas :
- create a reporter that fit to a existing file format already known by defectDojo
- create a tool for defectDojo to parse an ort evaluator result file

**Comment by sschuberth** [comment]
DefectDojo has a [REST API](https://defectdojo.readthedocs.io/en/latest/api-v2-docs.html) so we should probably realize this as an *Advisor*.

**Comment by fb33** [comment]
I need a integration with DefectDojo, so I'll do some code.
My idea is to generate a json report to export only unresolved rule violations from the ORT side.
This report will be published in DefectDojo threw API /import_scan
And on the DefectDojo side, I'm going to create an importer that will be called by the API and that will convert each rule violation into a defectdojo finding.

@sschuberth what do you think about that ?

**Comment by fb33** [comment]
The json report could look like that :
( based on reporter/src/funTest/assets/static-html-reporter-test-input.yml )
```
{
  "auditDate": "2020-10-26T15:30:20.491613800Z",
  "projects": [
    {
      "id": "Gradle:org.ossreviewtoolkit.gradle.example:lib:1.0.0",
      "rule_violations": [
        {
          "rule": "rule 1",
          "package": "Ant:junit:junit:4.12",
          "license": "EPL-1.0",
          "violationMessage": "EPL-1.0 error",
          "severity": "ERROR"
        },
        {
          "rule": "rule 3",
          "package": "Maven:org.hamcrest:hamcrest-core:1.3",
          "license": "BSD-3-Clause",
          "violationMessage": "BSD-3-Clause warning",
          "severity": "WARNING"
        }
      ]
    },
    {
      "id": "Gradle:org.ossreviewtoolkit:nested-fake-project:1.0.0",
      "rule_violations": [
        {
          "rule": "rule 1",
          "package": "Ant:junit:junit:4.12",
          "license": "EPL-1.0",
          "violationMessage": "EPL-1.0 error",
          "severity": "ERROR"
        },
        {
          "rule": "rule 3",
          "package": "Maven:org.hamcrest:hamcrest-core:1.3",
          "license": "BSD-3-Clause",
          "violationMessage": "BSD-3-Clause warning",
          "severity": "WARNING"
        }
      ]
    }
  ]
}
```

**Comment by sschuberth** [comment]
Interesting approach. I haven't used DefectDojo myself so far (and I had some misunderstandings about its scope / purpose, see my edited post above), so I cannot really give good advice here. But I'm wondering why an importer is required on the DefectDojo side (as I don't have access to the `import_scan` API docs). Does that API take any custom JSON, and that's why you need an importer at the DefectDojo side? Can't you just use another API that already expects a structure known to DefectDojo and generate that structure from ORT?

**Comment by fb33** [comment]
Hello, there's a "generic findings import" that can be used when calling /import_scan.
It parses a csv file.
So I can develop a defectDojoCsvReporter for ORT, that will map a unresolved rule violation in a defectdojo csv row.

**Comment by sschuberth** [comment]
The reason I'm asking is to minimize implementation effort. If possible, I'd try to avoid writing some custom exporter *and* an importer on the DefectDojo side. So, would is also be possible to let ORT's evaluator simply write an ORT result file (including rule violations) in JSON format, upload that via `import_scan` to DefectDojo, and *only* do some custom import / post-processing / filtering on the DefectDojo side? It would be awesome if DefectDojo could be made to understand standard ORT result files this way.

**Comment by fb33** [comment]
I understand your point of view, and I also think that this is the right way to do this integration in theory.
But several points make me think that it will be easier to write a specific reporter. Let me explain:
The ORT model contains a lot of information, so report files can be very large (they often exceed 100MB in my projects). Calling a Rest API with more than 100MB of payload is not necessarily a good thing.
The ORT model is complex and we have to develop and maintain in duplicate (kotlin and python) a logic of manipulation and interpretation of this model. I did not find any report allowing "simple" access to the unresolved rule_violations. Apparently, only the "evaluated-model.json" report contains all the information, but we have to reconcile the rule_violation_resolution and the rule_violations in order to deduce the unresolved ones.

I'm going to do a test to see if I can quickly and easily interpret the evaluated-model in Python.

**Comment by sschuberth** [comment]
> I'm going to do a test to see if I can quickly and easily interpret the evaluated-model in Python.

That sounds like a good plan!

I get your point about ORT result files being (too) large. So maybe we could also do it exactly the opposite: Write an ORT reporter for one of the 22+ formats (mentioned [here](https://www.defectdojo.org/#features)) that DefectDojo can *already* import, so no custom importer would need to be written on the DefectDojo side, but *only* the ORT reporter.

**Comment by fb33** [comment]
So, I've got some Python that maps unresolved rule violation to a "Finding".
[parser.py.txt](https://github.com/oss-review-toolkit/ort/files/5450920/parser.py.txt)
It allows to send the EVALUATEDMODELJSON report to `/import_scan` with this specific importer.

An other solution is to write a reporter to generate a defectdojo csv report, that can be send to `/import_scan` with this defectdojo generic importer (an already build in importer)

**Comment by sschuberth** [comment]
To me the decision which approach to take to import the data to DetectDojo is based on the question which approach allows to maintain the most data / map the data in the most meaningful way. If, say, the semantics of the predefined columns in the DefectDojo CSV do not fit well to the data provided by ORT, then I'd certainly prefer the approach you have taken to write a custom importer. That way we are not dependent on DefectDojo to extend their CSV. And as soon as maybe the internal data model of DefectDojo is extended, we could make use of that in the importer.

So, bottom line, I like the approach to take the existing evaluated model report and post-process it via a custom importer.

Do you plan to contribute your parser back to DefectDojo itself for inclusion?

**Comment by fb33** [comment]
Yes, that's the idea. 
But first I would have liked to see with you if the mapping corresponded to what you imagined.

**Comment by sschuberth** [comment]
That's really hard to tell for me as I have no experience with DefectDojo whatsoever, I don't have access a DefectDojo instance, and I do not know it's data model. And I lack time. I probably would have simply looked at how the imported data looks in the DefectDojo UI, and if it looks "good" / complete / meaningful.

**Comment by fb33** [comment]
Hello, for your information, I made a PR in DefectDojo 
https://github.com/DefectDojo/django-DefectDojo/pull/3177

**Comment by sschuberth** [comment]
@fb33 as your PR was merged in DefectDojo, can this be closed?

**Comment by fb33** [comment]
Yes, I close it.
